enhancements/authentication/login-logout-events.md

```diff
@@ -0,0 +1,330 @@
+2. Extend oauth-server to capture login and login failure events.
+3. Extend oauth-apiserver to add user information when deleting `oauthaccesstoken` objects.
+4. Extend must-gather/gather_audit_logs to fetch oauth-server audit logs.
+
+### Non-Goals
+
+1. Hook into the existing API server [policy configuration](https://github.com/openshift/enhancements/blob/master/enhancements/kube-apiserver/audit-policy.md).
+2. Provide a fully featured audit subsystem in oauth-server compliant with API server policy framework.
+3. Rate limiting of logging login/logout events.
+
+## Proposal
+
+We propose to add an `audit` struct to `oauth.config.openshift.io/v1` with a `profile` field which configures audit logging granularity.
+
+```go
+// OAuthSpec contains desired cluster auth configuration
+type OAuthSpec struct {
+	// identityProviders is an ordered list of ways for a user to identify themselves.
+	// When this list is empty, no identities are provisioned for users.
+	// +optional
+	// +listType=atomic
+	IdentityProviders []IdentityProvider `json:"identityProviders,omitempty"`
+
+	// tokenConfig contains options for authorization and access tokens
+	TokenConfig TokenConfig `json:"tokenConfig"`
+
+	// templates allow you to customize pages like the login page.
+	// +optional
+	Templates OAuthTemplates `json:"templates"`
+	// audit specifies what should be audited in the context of OAuthServer.
+	// +optional
+	// +kubebuilder:default:={"profile":"WriteLoginEvents"}
+	Audit OAuthAudit `json:"audit"`
+}
+
+// OAuthAudit specifies the Audit profile in use.
+type OAuthAudit struct {
+	// profile is a simple drop in profile type that can be turned off by
+	// setting it to "None" or it can be turned on by setting it to
+	// "WriteLoginEvents".
+	// +kubebuilder:default:="WriteLoginEvents"
+	Profile OAuthAuditProfileType `json:"profile,omitempty"`
+}
+
+// OAuthAuditProfileType defines a simple audit profile, which can turn OAuth
+// authentication audit logging on or off.
+// +kubebuilder:validation:Enum=None;WriteLoginEvents
+type OAuthAuditProfileType string
+
+const (
+	// "None" disables audit logs.
+	OAuthNoneAuditProfileType AuditProfileType = "None"
+
+	// "WriteLoginEvents" logs login and login failure events.
+	// This is the default.
+	OAuthWriteLoginEventsProfileType AuditProfileType = "WriteLoginEvents"
+)
+```
+
+### User Stories
+
+#### As an administrator I want to inspect successful login and login failure attempts
+
+Use `oc adm must-gather -- /usr/bin/gather_audit_logs` and inspect oauth-server audit logs.
+
+#### As an administrator I want to inspect logout events and the deleted user
+
+Use `oc adm must-gather -- /usr/bin/gather_audit_logs` and inspect apiserver audit logs for deleted `oauthaccesstoken` resources.
+Audit log entries include the deleted user.
+
+#### As an administrator I want to inspect the source IPs of the client triggering an authentication event
+
```

Comment:
please add how this use case is implemented. namely: Using `oc adm must-gather -- /usr/bin/gather_audit_logs` and inspecting oauth-server audit logs one can see the source IP of the client.